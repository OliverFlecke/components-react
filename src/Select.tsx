import React, { forwardRef, useCallback } from 'react';

export interface SelectProps {
	label: string;
	children: any[];
	onSelection?: (value: string) => void;
}

const Select = forwardRef<
	HTMLSelectElement,
	SelectProps & React.SelectHTMLAttributes<HTMLSelectElement>
>(({ label, children, onSelection, ...selectProps }, ref) => {
	const changed = useCallback(
		(e: React.ChangeEvent<HTMLSelectElement>) => {
			onSelection?.(e.currentTarget.value);
		},
		[onSelection]
	);

	return (
		<label className="flex flex-col space-y-2 text-black dark:text-gray-100">
			<span className="input-label">{label}</span>
			<select
				ref={ref}
				className="rounded-md px-4 py-2 bg-gray-100 dark:bg-gray-900 dark:text-white shadow focus:outline-none focus:ring focus:border-indigo-400"
				onChange={changed}
				{...selectProps}
			>
				{children}
			</select>
		</label>
	);
});

interface SelectOptionProps {
	value: any;
	children?: string;
}

export const SelectOption = ({ value, children }: SelectOptionProps) => (
	<option value={value}>{children ?? value}</option>
);

interface SelectGroupProps {
	label: string;
	children: any[];
}

export const SelectGroup = ({ children, label }: SelectGroupProps) => (
	<optgroup label={label}>{children}</optgroup>
);

export default Select as any;
