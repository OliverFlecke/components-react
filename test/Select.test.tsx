import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import React from 'react';
import renderer from 'react-test-renderer';
import Select, { SelectGroup, SelectOption } from '../src/Select';

Enzyme.configure({ adapter: new Adapter() });

describe('Select', () => {
	test('Render snapshot', () => {
		const tree = renderer
			.create(
				<Select>
					<SelectGroup label="values">
						<SelectOption value="1">1</SelectOption>
						<SelectOption value="2">2</SelectOption>
					</SelectGroup>
				</Select>
			)
			.toJSON();
		expect(tree).toMatchSnapshot();
	});
});
