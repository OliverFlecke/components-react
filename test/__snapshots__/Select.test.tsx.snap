// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select Render snapshot 1`] = `
<label
  className="flex flex-col space-y-2"
>
  <span
    className="input-label"
  />
  <select
    className="rounded-md px-4 py-2 bg-gray-100 dark:bg-gray-900 dark:text-white shadow focus:outline-none focus:ring focus:border-indigo-400"
    onChange={[Function]}
  >
    <optgroup
      label="values"
    >
      <option
        value="1"
      >
        1
      </option>
      <option
        value="2"
      >
        2
      </option>
    </optgroup>
  </select>
</label>
`;
